File: go/adbc_test.go
package main

import (
	"context"
	"fmt"
	"os"
	"testing"
	"time"

	"github.com/apache/arrow-adbc/go/adbc"
	"github.com/apache/arrow-adbc/go/adbc/driver/flightsql"
	"github.com/apache/arrow/go/v17/arrow/memory"
)

func getFlightSqlURI() string {
	host := os.Getenv("XTDB_HOST")
	if host == "" {
		host = "xtdb"
	}
	return fmt.Sprintf("grpc://%s:9833", host)
}

var adbcTableCounter int

func getAdbcCleanTable() string {
	adbcTableCounter++
	return fmt.Sprintf("test_adbc_%d_%d", time.Now().Unix(), adbcTableCounter)
}

// Helper to create an ADBC connection
func getAdbcConn(t *testing.T) (adbc.Database, adbc.Connection) {
	alloc := memory.NewGoAllocator()
	driver := flightsql.NewDriver(alloc)

	db, err := driver.NewDatabase(map[string]string{
		"uri": getFlightSqlURI(),
	})
	if err != nil {
		t.Fatalf("Failed to create database: %v", err)
	}

	conn, err := db.Open(context.Background())
	if err != nil {
		db.Close()
		t.Fatalf("Failed to open connection: %v", err)
	}

	return db, conn
}

func cleanupAdbc(conn adbc.Connection, table string, ids ...int) {
	ctx := context.Background()
	for _, id := range ids {
		stmt, err := conn.NewStatement()
		if err != nil {
			continue
		}
		stmt.SetSqlQuery(fmt.Sprintf("ERASE FROM %s WHERE _id = %d", table, id))
		stmt.ExecuteUpdate(ctx)
		stmt.Close()
	}
}

// === Connection Tests ===

func TestAdbcConnection(t *testing.T) {
	db, conn := getAdbcConn(t)
	defer conn.Close()
	defer db.Close()

	if conn == nil {
		t.Fatal("Connection should be established")
	}
}

func TestAdbcSimpleQuery(t *testing.T) {
	db, conn := getAdbcConn(t)
	defer conn.Close()
	defer db.Close()

	ctx := context.Background()
	stmt, err := conn.NewStatement()
	if err != nil {
		t.Fatalf("Failed to create statement: %v", err)
	}
	defer stmt.Close()

	stmt.SetSqlQuery("SELECT 1 AS x, 'hello' AS greeting")
	reader, _, err := stmt.ExecuteQuery(ctx)
	if err != nil {
		t.Fatalf("Query failed: %v", err)
	}
	defer reader.Release()

	if !reader.Next() {
		t.Fatal("Should have at least one batch")
	}

	record := reader.Record()
	if record.NumRows() != 1 {
		t.Errorf("Expected 1 row, got %d", record.NumRows())
	}
	if record.NumCols() != 2 {
		t.Errorf("Expected 2 columns, got %d", record.NumCols())
	}
}

func TestAdbcQueryWithExpressions(t *testing.T) {
	db, conn := getAdbcConn(t)
	defer conn.Close()
	defer db.Close()

	ctx := context.Background()
	stmt, err := conn.NewStatement()
	if err != nil {
		t.Fatalf("Failed to create statement: %v", err)
	}
	defer stmt.Close()

	stmt.SetSqlQuery("SELECT 2 + 2 AS sum, UPPER('hello') AS upper_greeting")
	reader, _, err := stmt.ExecuteQuery(ctx)
	if err != nil {
		t.Fatalf("Query failed: %v", err)
	}
	defer reader.Release()

	if !reader.Next() {
		t.Fatal("Should have at least one batch")
	}

	record := reader.Record()
	if record.NumRows() != 1 {
		t.Errorf("Expected 1 row, got %d", record.NumRows())
	}
}

func TestAdbcSystemTables(t *testing.T) {
	db, conn := getAdbcConn(t)
	defer conn.Close()
	defer db.Close()

	ctx := context.Background()
	stmt, err := conn.NewStatement()
	if err != nil {
		t.Fatalf("Failed to create statement: %v", err)
	}
	defer stmt.Close()

	stmt.SetSqlQuery("SELECT table_name FROM information_schema.tables WHERE table_schema = 'public' LIMIT 10")
	reader, _, err := stmt.ExecuteQuery(ctx)
	if err != nil {
		t.Fatalf("Query failed: %v", err)
	}
	defer reader.Release()

	// Should execute without error
	if reader == nil {
		t.Fatal("Reader should not be nil")
	}
}

// === DML Tests ===

func TestAdbcInsertAndQuery(t *testing.T) {
	db, conn := getAdbcConn(t)
	defer conn.Close()
	defer db.Close()

	ctx := context.Background()
	table := getAdbcCleanTable()

	// INSERT using RECORDS syntax
	stmt, err := conn.NewStatement()
	if err != nil {
		t.Fatalf("Failed to create statement: %v", err)
	}

	stmt.SetSqlQuery(fmt.Sprintf(
		"INSERT INTO %s RECORDS "+
			"{_id: 1, name: 'Widget', price: 19.99, category: 'gadgets'}, "+
			"{_id: 2, name: 'Gizmo', price: 29.99, category: 'gadgets'}, "+
			"{_id: 3, name: 'Thingamajig', price: 9.99, category: 'misc'}",
		table,
	))
	_, err = stmt.ExecuteUpdate(ctx)
	stmt.Close()
	if err != nil {
		t.Fatalf("Insert failed: %v", err)
	}


	// Query the inserted data
	stmt2, err := conn.NewStatement()
	if err != nil {
		t.Fatalf("Failed to create statement: %v", err)
	}
	defer stmt2.Close()

	stmt2.SetSqlQuery(fmt.Sprintf("SELECT * FROM %s ORDER BY _id", table))
	reader, _, err := stmt2.ExecuteQuery(ctx)
	if err != nil {
		t.Fatalf("Query failed: %v", err)
	}
	defer reader.Release()

	if !reader.Next() {
		t.Fatal("Should have at least one batch")
	}

	record := reader.Record()
	if record.NumRows() != 3 {
		t.Errorf("Expected 3 rows, got %d", record.NumRows())
	}

	// Cleanup
	cleanupAdbc(conn, table, 1, 2, 3)
}

func TestAdbcUpdate(t *testing.T) {
	db, conn := getAdbcConn(t)
	defer conn.Close()
	defer db.Close()

	ctx := context.Background()
	table := getAdbcCleanTable()

	// Insert initial data
	stmt, _ := conn.NewStatement()
	stmt.SetSqlQuery(fmt.Sprintf("INSERT INTO %s RECORDS {_id: 1, name: 'Widget', price: 19.99}", table))
	stmt.ExecuteUpdate(ctx)
	stmt.Close()

	// Update the price
	stmt2, _ := conn.NewStatement()
	stmt2.SetSqlQuery(fmt.Sprintf("UPDATE %s SET price = 24.99 WHERE _id = 1", table))
	stmt2.ExecuteUpdate(ctx)
	stmt2.Close()

	// Verify update
	stmt3, _ := conn.NewStatement()
	defer stmt3.Close()
	stmt3.SetSqlQuery(fmt.Sprintf("SELECT price FROM %s WHERE _id = 1", table))
	reader, _, err := stmt3.ExecuteQuery(ctx)
	if err != nil {
		t.Fatalf("Query failed: %v", err)
	}
	defer reader.Release()

	if !reader.Next() {
		t.Fatal("Should have at least one batch")
	}

	record := reader.Record()
	if record.NumRows() != 1 {
		t.Errorf("Expected 1 row, got %d", record.NumRows())
	}

	// Cleanup
	cleanupAdbc(conn, table, 1)
}

func TestAdbcDelete(t *testing.T) {
	db, conn := getAdbcConn(t)
	defer conn.Close()
	defer db.Close()

	ctx := context.Background()
	table := getAdbcCleanTable()

	// Insert data
	stmt, _ := conn.NewStatement()
	stmt.SetSqlQuery(fmt.Sprintf("INSERT INTO %s RECORDS {_id: 1, name: 'ToDelete'}, {_id: 2, name: 'ToKeep'}", table))
	stmt.ExecuteUpdate(ctx)
	stmt.Close()

	// Delete one record
	stmt2, _ := conn.NewStatement()
	stmt2.SetSqlQuery(fmt.Sprintf("DELETE FROM %s WHERE _id = 1", table))
	stmt2.ExecuteUpdate(ctx)
	stmt2.Close()

	// Verify only one record remains
	stmt3, _ := conn.NewStatement()
	defer stmt3.Close()
	stmt3.SetSqlQuery(fmt.Sprintf("SELECT * FROM %s", table))
	reader, _, err := stmt3.ExecuteQuery(ctx)
	if err != nil {
		t.Fatalf("Query failed: %v", err)
	}
	defer reader.Release()

	if !reader.Next() {
		t.Fatal("Should have at least one batch")
	}

	record := reader.Record()
	if record.NumRows() != 1 {
		t.Errorf("Expected 1 row, got %d", record.NumRows())
	}

	// Cleanup
	cleanupAdbc(conn, table, 1, 2)
}

func TestAdbcHistoricalQuery(t *testing.T) {
	db, conn := getAdbcConn(t)
	defer conn.Close()
	defer db.Close()

	ctx := context.Background()
	table := getAdbcCleanTable()

	// Insert initial data
	stmt, _ := conn.NewStatement()
	stmt.SetSqlQuery(fmt.Sprintf("INSERT INTO %s RECORDS {_id: 1, name: 'Widget', price: 19.99}", table))
	stmt.ExecuteUpdate(ctx)
	stmt.Close()

	// Update (creates new version)
	stmt2, _ := conn.NewStatement()
	stmt2.SetSqlQuery(fmt.Sprintf("UPDATE %s SET price = 24.99 WHERE _id = 1", table))
	stmt2.ExecuteUpdate(ctx)
	stmt2.Close()

	// Query historical data
	stmt3, _ := conn.NewStatement()
	defer stmt3.Close()
	stmt3.SetSqlQuery(fmt.Sprintf(
		"SELECT *, _valid_from, _valid_to FROM %s FOR ALL VALID_TIME ORDER BY _id, _valid_from",
		table,
	))
	reader, _, err := stmt3.ExecuteQuery(ctx)
	if err != nil {
		t.Fatalf("Query failed: %v", err)
	}
	defer reader.Release()

	if !reader.Next() {
		t.Fatal("Should have at least one batch")
	}

	record := reader.Record()
	// Should have 2 versions
	if record.NumRows() != 2 {
		t.Errorf("Expected 2 rows (historical versions), got %d", record.NumRows())
	}

	// Cleanup
	cleanupAdbc(conn, table, 1)
}

func TestAdbcErase(t *testing.T) {
	db, conn := getAdbcConn(t)
	defer conn.Close()
	defer db.Close()

	ctx := context.Background()
	table := getAdbcCleanTable()

	// Insert data
	stmt, _ := conn.NewStatement()
	stmt.SetSqlQuery(fmt.Sprintf("INSERT INTO %s RECORDS {_id: 1, name: 'ToErase'}, {_id: 2, name: 'ToKeep'}", table))
	stmt.ExecuteUpdate(ctx)
	stmt.Close()

	// Update to create history
	stmt2, _ := conn.NewStatement()
	stmt2.SetSqlQuery(fmt.Sprintf("UPDATE %s SET name = 'UpdatedErase' WHERE _id = 1", table))
	stmt2.ExecuteUpdate(ctx)
	stmt2.Close()

	// Erase record 1 completely
	stmt3, _ := conn.NewStatement()
	stmt3.SetSqlQuery(fmt.Sprintf("ERASE FROM %s WHERE _id = 1", table))
	stmt3.ExecuteUpdate(ctx)
	stmt3.Close()

	// Verify erased from all history
	stmt4, _ := conn.NewStatement()
	defer stmt4.Close()
	stmt4.SetSqlQuery(fmt.Sprintf("SELECT * FROM %s FOR ALL VALID_TIME ORDER BY _id", table))
	reader, _, err := stmt4.ExecuteQuery(ctx)
	if err != nil {
		t.Fatalf("Query failed: %v", err)
	}
	defer reader.Release()

	if !reader.Next() {
		t.Fatal("Should have at least one batch")
	}

	record := reader.Record()
	// Only record 2 should remain
	if record.NumRows() != 1 {
		t.Errorf("Expected 1 row, got %d", record.NumRows())
	}

	// Cleanup
	cleanupAdbc(conn, table, 2)
}

File: csharp/AdbcTest.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Apache.Arrow;
using Apache.Arrow.Adbc;
using Apache.Arrow.Adbc.Drivers.Interop.FlightSql;
using Xunit;

namespace XtdbTests;

/// <summary>
/// XTDB ADBC Tests
///
/// Tests for connecting to XTDB via Arrow Flight SQL protocol using ADBC.
/// Uses the Interop driver which wraps the mature Go Flight SQL implementation.
/// Demonstrates DML operations (INSERT, UPDATE, DELETE, ERASE) and temporal queries.
/// </summary>
public class AdbcTest : IDisposable
{
    private static readonly string FlightSqlUri = GetFlightSqlUri();
    private static int _tableCounter = 0;

    private static string GetFlightSqlUri()
    {
        var host = Environment.GetEnvironmentVariable("XTDB_HOST") ?? "xtdb";
        return $"grpc://{host}:9833";
    }

    private readonly AdbcDriver _driver;
    private readonly AdbcDatabase _database;
    private readonly AdbcConnection _connection;

    public AdbcTest()
    {
        // Load the Go-based Flight SQL driver via interop
        _driver = FlightSqlDriverLoader.LoadDriver();

        var parameters = new Dictionary<string, string>
        {
            { "uri", FlightSqlUri }
        };
        _database = _driver.Open(parameters);
        // Connect() takes no parameters for the Interop driver
        _connection = _database.Connect(new Dictionary<string, string>());
    }

    public void Dispose()
    {
        _connection?.Dispose();
        _database?.Dispose();
    }

    private static string GetCleanTable()
    {
        Interlocked.Increment(ref _tableCounter);
        return $"test_adbc_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_{_tableCounter}";
    }

    private void Cleanup(string table, params int[] ids)
    {
        foreach (var id in ids)
        {
            try
            {
                using var stmt = _connection.CreateStatement();
                stmt.SqlQuery = $"ERASE FROM {table} WHERE _id = {id}";
                stmt.ExecuteUpdate();
            }
            catch
            {
                // Ignore cleanup errors
            }
        }
    }

    // === Connection Tests ===

    [Fact]
    public void TestConnection()
    {
        Assert.NotNull(_connection);
    }

    [Fact]
    public void TestSimpleQuery()
    {
        using var stmt = _connection.CreateStatement();
        stmt.SqlQuery = "SELECT 1 AS x, 'hello' AS greeting";

        var result = stmt.ExecuteQuery();
        Assert.NotNull(result.Stream);

        var batch = result.Stream.ReadNextRecordBatchAsync().Result;
        Assert.NotNull(batch);
        Assert.Equal(1, batch.Length);
        Assert.Equal(2, batch.ColumnCount);

        // Verify column names
        var schema = batch.Schema;
        var columnNames = new List<string>();
        foreach (var field in schema.FieldsList)
        {
            columnNames.Add(field.Name);
        }
        Assert.Contains("x", columnNames);
        Assert.Contains("greeting", columnNames);
    }

    [Fact]
    public void TestQueryWithExpressions()
    {
        using var stmt = _connection.CreateStatement();
        stmt.SqlQuery = "SELECT 2 + 2 AS sum, UPPER('hello') AS upper_greeting";

        var result = stmt.ExecuteQuery();
        var batch = result.Stream.ReadNextRecordBatchAsync().Result;

        Assert.NotNull(batch);
        Assert.Equal(1, batch.Length);
    }

    [Fact]
    public void TestSystemTables()
    {
        using var stmt = _connection.CreateStatement();
        stmt.SqlQuery = "SELECT table_name FROM information_schema.tables WHERE table_schema = 'public' LIMIT 10";

        var result = stmt.ExecuteQuery();
        Assert.NotNull(result.Stream);
    }

    // === DML Tests ===

    [Fact]
    public void TestInsertAndQuery()
    {
        var table = GetCleanTable();

        try
        {
            // INSERT using RECORDS syntax
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"INSERT INTO {table} RECORDS " +
                    "{_id: 1, name: 'Widget', price: 19.99, category: 'gadgets'}, " +
                    "{_id: 2, name: 'Gizmo', price: 29.99, category: 'gadgets'}, " +
                    "{_id: 3, name: 'Thingamajig', price: 9.99, category: 'misc'}";
                stmt.ExecuteUpdate();
            }


            // Query the inserted data
            using var queryStmt = _connection.CreateStatement();
            queryStmt.SqlQuery = $"SELECT * FROM {table} ORDER BY _id";
            var result = queryStmt.ExecuteQuery();
            var batch = result.Stream.ReadNextRecordBatchAsync().Result;

            Assert.NotNull(batch);
            Assert.Equal(3, batch.Length);
        }
        finally
        {
            Cleanup(table, 1, 2, 3);
        }
    }

    [Fact]
    public void TestUpdate()
    {
        var table = GetCleanTable();

        try
        {
            // Insert initial data
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"INSERT INTO {table} RECORDS {{_id: 1, name: 'Widget', price: 19.99}}";
                stmt.ExecuteUpdate();
            }

            // Update the price
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"UPDATE {table} SET price = 24.99 WHERE _id = 1";
                stmt.ExecuteUpdate();
            }

            // Verify update
            using var queryStmt = _connection.CreateStatement();
            queryStmt.SqlQuery = $"SELECT price FROM {table} WHERE _id = 1";
            var result = queryStmt.ExecuteQuery();
            var batch = result.Stream.ReadNextRecordBatchAsync().Result;

            Assert.NotNull(batch);
            Assert.Equal(1, batch.Length);
        }
        finally
        {
            Cleanup(table, 1);
        }
    }

    [Fact]
    public void TestDelete()
    {
        var table = GetCleanTable();

        try
        {
            // Insert data
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"INSERT INTO {table} RECORDS {{_id: 1, name: 'ToDelete'}}, {{_id: 2, name: 'ToKeep'}}";
                stmt.ExecuteUpdate();
            }

            // Delete one record
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"DELETE FROM {table} WHERE _id = 1";
                stmt.ExecuteUpdate();
            }

            // Verify only one record remains
            using var queryStmt = _connection.CreateStatement();
            queryStmt.SqlQuery = $"SELECT * FROM {table}";
            var result = queryStmt.ExecuteQuery();
            var batch = result.Stream.ReadNextRecordBatchAsync().Result;

            Assert.NotNull(batch);
            Assert.Equal(1, batch.Length);
        }
        finally
        {
            Cleanup(table, 1, 2);
        }
    }

    [Fact]
    public void TestHistoricalQuery()
    {
        var table = GetCleanTable();

        try
        {
            // Insert initial data
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"INSERT INTO {table} RECORDS {{_id: 1, name: 'Widget', price: 19.99}}";
                stmt.ExecuteUpdate();
            }

            // Update (creates new version)
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"UPDATE {table} SET price = 24.99 WHERE _id = 1";
                stmt.ExecuteUpdate();
            }

            // Query historical data
            using var queryStmt = _connection.CreateStatement();
            queryStmt.SqlQuery = $"SELECT *, _valid_from, _valid_to FROM {table} FOR ALL VALID_TIME ORDER BY _id, _valid_from";
            var result = queryStmt.ExecuteQuery();
            var batch = result.Stream.ReadNextRecordBatchAsync().Result;

            Assert.NotNull(batch);
            // Should have 2 versions
            Assert.Equal(2, batch.Length);
        }
        finally
        {
            Cleanup(table, 1);
        }
    }

    [Fact]
    public void TestErase()
    {
        var table = GetCleanTable();

        try
        {
            // Insert data
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"INSERT INTO {table} RECORDS {{_id: 1, name: 'ToErase'}}, {{_id: 2, name: 'ToKeep'}}";
                stmt.ExecuteUpdate();
            }

            // Update to create history
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"UPDATE {table} SET name = 'UpdatedErase' WHERE _id = 1";
                stmt.ExecuteUpdate();
            }

            // Erase record 1 completely
            using (var stmt = _connection.CreateStatement())
            {
                stmt.SqlQuery = $"ERASE FROM {table} WHERE _id = 1";
                stmt.ExecuteUpdate();
            }

            // Verify erased from all history
            using var queryStmt = _connection.CreateStatement();
            queryStmt.SqlQuery = $"SELECT * FROM {table} FOR ALL VALID_TIME ORDER BY _id";
            var result = queryStmt.ExecuteQuery();
            var batch = result.Stream.ReadNextRecordBatchAsync().Result;

            Assert.NotNull(batch);
            // Only record 2 should remain
            Assert.Equal(1, batch.Length);
        }
        finally
        {
            Cleanup(table, 2);
        }
    }
}
